programs/vkd3d-hlsl-build/main.cpp

```diff
@@ -0,0 +1,478 @@
+    /* Full target name for d3dcompiler */
+    std::array<char, 16> name = { };
+    /* Whether or not the shader targets Shader Model 6 or above */
+    bool is_dxil = false;
+
+    template<typename char_type>
+    bool parse(const char_type *str)
+    {
+        uint32_t idx = 0;
+
+        if (str[idx++] != char_type('.'))
+            return false;
+
+        /* Validate profile name until delimiter. Only ASCII
+         * characters are allowed. */
+        while (str[idx] && str[idx] != char_type('_'))
+        {
+            if (str[idx++] > char_type('\x7f'))
+                return false;
+        }
+
+        if (str[idx++] != char_type('_'))
+            return false;
+
+        /* Valicate major version and check whether the profile is
+         * DXIL (6_x) or DXBC (5_x) */
+        if (str[idx] < char_type('0') || str[idx] > char_type('9'))
+            return false;
+
+        uint8_t hi_ver = uint8_t(str[idx++] - char_type('0'));
+
+        if (str[idx++] != char_type('_'))
+            return false;
+
+        is_dxil = hi_ver >= 6;
+
+        /* Validate minor version number */
+        while (str[idx])
+        {
+            if (str[idx] < char_type('0') || str[idx] > char_type('9'))
+                return false;
+            idx++;
+        }
+
+        /* Make sure we can use the fixed-size name array */
+        if (idx > name.size())
+            return false;
+
+        /* Skip initial dot when converting strings */
+        for (uint32_t i = 1; i < idx; i++)
+        {
+            wide_name[i - 1] = str[i];
+            name[i - 1] = char(str[i]);
+        }
+
+        return true;
+    }
+};
+
+
+struct work_item
+{
+    std::filesystem::path in_file;
+    std::filesystem::path out_file;
+
+    /* Support one DXIL profile and one DXBC one */
+    std::array<shader_profile, 2> profiles;
+    uint32_t profile_count = 0;
+
+
+    work_item(std::filesystem::path file, const std::filesystem::path &dst_dir)
+    : in_file(std::move(file))
+    {
+        /* Get rid of .hlsl extension right away */
+        auto file_name = in_file.filename().stem();
+
+        /* Parse the file name as a sequence of shader targets delimited
+         * by dots, e.g. frog.vs_6_6.vs_5_1.hlsl */
+        while (true)
+        {
+            auto extension = file_name.extension();
+
+            if (extension.empty())
+            {
+                /* Only the base file name is left at this point, use it as
+                 * the target file name as well as the variable name within
+                 * the generated header. */
+                out_file = dst_dir / file_name;
+                out_file += ".h";
+                break;
+            }
+
+            if (profile_count == profiles.size())
+            {
+                std::wcerr << in_file << ": Too many profiles." << std::endl;
+                break;
+            }
+
+            if (!profiles.at(profile_count).parse(extension.c_str()))
```

Comment:
The critical part here is that we need a way to pass down custom command line options. E.g. some of the shaders need -enable-16bit-types, ftz denorm modes, etc.

I'm not too picky which direction we choose. SPIRV-Cross and dxil-spirv use a style where we have:

foo.dothisthing.dothatthing.frag

and the python script looks for pre-determined patterns of `.dothisthing.` and appends CLI arguments as necessary.

Another idea would be to make this completely generic where we can tag with e.g. `frog.native-types.cs_6_2.hlsl` and CLI can look up the file `native-types.txt` or something where the command line is written. This tool seems useful enough in general that it might be worth going the generic route.

Reply:
I changed this to interpret everything that isn't a shader profile as a command line option, e.g. `frog.vs_6_0.denorm ftz.hlsl` would be equivalent to passing `-denorm ftz` to DXC.